### PR TITLE
ARROW-17883: [Java] implement immutable table

### PR DESCRIPTION
**Pull request by lwhite1**
Table is a new immutable tabular data structure based on FieldVectors.

This PR is described in detail in the included README.md file. The original design discussion can be found [here](https://docs.google.com/document/d/1J77irZFWNnSID7vK71z26Nw_Pi99I9Hb9iryno8B03c/edit#heading=h.a1lebwljypq5), if you're interested.

Note to reviewers:
- This is a fairly large change set. Most of the code is in "getters" in the Row class. These methods are fairly well covered by tests, but it would be good to have someone look especially at the complex vector types. 
- The only changes to existing classes were three new export methods added to the Data class. These use the logic for exporting VectorSchemaRoots.

### CODE REVIEW

**Comment by lwhite1** [comment]
If @lidavidm and @davisusanibar could do a review it would be greatly appreciated.

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+Since rows are iterable, you can traverse a table using a standard while loop:
+
+```java
+Row r = table.immutableRow();
+    while (c.hasNext()) {
+    c.next();
+    // do something useful here
+    }
+```
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/table/README.md:270]: Should `c` here be `r`?

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+    int b = row.getInt("age", holder);
+```
+
+This can be used to retrieve values without creating a new Object for each.
+
+In addition to getting values, you can check if a value is null using `isNull()` and you can get the current row number:
+
+```java
+boolean name0isNull = row.isNull("name");
+    int row = row.getRowNumber();
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/table/README.md:317]: It may be good to use `long` instead of `int` if we do anticipate adding ChunkedArray support in the future

File: java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Provides row based access to the data held by a {@link Table}.
+ *
+ * <p>
+ * This API is EXPERIMENTAL.
+ */
+public abstract class BaseRow {
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java:29]: For the class hierarchy: is a hypothetical MutableRow expected to inherit from BaseRow (in which case there's not a great way to abstract over immutable/mutable rows for read-only access) or from Row (in which case BaseRow seems a little redundant)?

File: java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  /** The table we're enumerating. */
+  protected final BaseTable table;
+
+  /** the current row number. */
+  protected int rowNumber = -1;
+
+  /**
+   * Returns the standard character set to use for decoding strings. The Arrow format only supports UTF-8.
+   */
+  private final Charset defaultCharacterSet = StandardCharsets.UTF_8;
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java:40]: It seems this should be a static, since UTF-8 is part of the standard. (I could see there being a non-final attribute for the case of getting strings out of a binary vector, though.)

File: java/vector/src/main/java/org/apache/arrow/vector/table/BaseTable.java [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+  protected DictionaryProvider dictionaryProvider;
+
+  /** A map of Fields to FieldVectors used to select Fields. */
+  protected final Map<Field, FieldVector> fieldVectorsMap = new LinkedHashMap<>();
+
+  /** The schema for the table. */
+  protected Schema schema;
+
+  /** The number of rows of data in the table; not necessarily the same as the table row capacity. */
+  protected int rowCount;
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/table/BaseTable.java:62]: Same here - I think it'd be best to use `long` everywhere for indexing from the get-go if we do anticipate ChunkedArray support (things currently use int because a single allocation can't hold more elements than that, but ChunkedArray would remove that restriction)

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+```Java
+Table t = new Table(vsr, provider);
+    ValueVector encodedName = t.encode("name", 1L);
+```
+
+***Current Limitation:*** One difference is the method that produces TSV formatted output has an extra switch instructing the Table to replace the encoded output with the decoded output where possible:
+
+```java
+String output = myTable.contentToTSVString(true);
+```
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/table/README.md:165]: It seems this doesn't exist in the code as written?

**Review by lidavidm — COMMENTED** [review]
Generally looks reasonable, I left some small comments.

Some other notes:
- The core code uses FreeMarker templating to help cut down on code duplication (particularly for Row.java), though it has its own issues
- I wonder if parameterized tests might help cut down some of the test code? I'm not sure if the Java library has enough metaprogramming capabilities for that to be helpful (the C++ libraries do)
- It might be good to port the README to Sphinx (at some point) so it shows up in the actual documentation

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md (hunk 1 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/README.md:270]: I will fix this.

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md (hunk 2 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/README.md:317]: Will do

File: java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java (hunk 4 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java:40]: Yes. I could change that. It was created as a field before I learned that only UTF-8 was supported.

File: java/vector/src/main/java/org/apache/arrow/vector/table/BaseTable.java (hunk 5 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/BaseTable.java:62]: Ok

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md (hunk 6 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/README.md:165]: True. I should remove that text until it gets implemented.

File: java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java (hunk 3 above)
Comment (lidavidm) [reply java/vector/src/main/java/org/apache/arrow/vector/table/BaseRow.java:29]: Since things are experimental, I think we can keep it for now and see how the mutable implementation actually looks

**Comment by lidavidm** [comment]
Ok, sounds good. Those comments aren't requirements, and if you already looked at FreeMarker then that's fine (it's hard to follow how it works, so I'm not a big fan of it unless it would reduce code size a lot). The main thing IMO is just being consistent about `long` vs `int`

**Comment by lwhite1** [comment]
> The main thing IMO is just being consistent about `long` vs `int`

I would like to walk back committing to that change now that I've looked at the code and thought about it. There are two issues, the first being relatively trivial, but the second being important I think:

1. There are many (over 100) places where an int is required in the current code. Most can be fixed with search-and-replace, but still...
2. If you allow longs now in the API, the user will get a runtime exception (array index out of bounds) if they go outside the int range due to the int value limit in ValueVector.  

Because of the second issue, I think it would be better to expand the range when longs are really supported than to expand it now and create a hazard in the code. It's a trivial API change later to go from int to long and no one will be adversely affected either then or now.

**Comment by lidavidm** [comment]
Ok. It'll affect binary compatibility but I don't think that's been a concern for us so far, so no worries (and this is experimental code anyways)

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md (hunk 2 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/README.md:316]: fixed here and elsewhere in the doc

File: java/vector/src/main/java/org/apache/arrow/vector/table/README.md (hunk 6 above)
Comment (lwhite1) [reply java/vector/src/main/java/org/apache/arrow/vector/table/README.md]: removed.

**Comment by lwhite1** [comment]
I removed the BaseRow class and converted the Charset variable to static.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
@davisusanibar any comments?

**Comment by lidavidm** [comment]
Looking at CI failures: the manylinux failure seems to be a SIGABRT - possibly there's a bug in the C++ side of the C Data Interface module. And the Windows failure appears to be a Windows code for "heap corruption" - not sure what we can do about that

**Comment by lwhite1** [comment]
> Looking at CI failures: the manylinux failure seems to be a SIGABRT - possibly there's a bug in the C++ side of the C Data Interface module. And the Windows failure appears to be a Windows code for "heap corruption" - not sure what we can do about that

Thanks for looking into that. It didn't seem like it could be related to the changes in the PR, but I wasn't sure.

**Comment by lidavidm** [comment]
If you want to copy paste them into Jiras - or I'll do it later (the Windows one seems iffy but I feel like I've seen it several times now) - the helpful thing would actually be to update CI to collect those core dump files when they happen so we can debug more easily

**Comment by lidavidm** [comment]
Regardless, not a blocker here

**Comment by davisusanibar** [comment]
> If you want to copy paste them into Jiras - or I'll do it later (the Windows one seems iffy but I feel like I've seen it several times now) - the helpful thing would actually be to update CI to collect those core dump files when they happen so we can debug more easily

- For Java JNI Gandiva errors, just created tis jira ticket: https://issues.apache.org/jira/browse/ARROW-17946
- Not able to replicate Windows issue locally, just created the issue with message on this PR: https://issues.apache.org/jira/browse/ARROW-17947

**Comment by lwhite1** [comment]
@davisusanibar Do you approve the PR?

**Review by davisusanibar — APPROVED** [review]
LGTM, thank you

**Comment by lwhite1** [comment]
> I only have this question: At current status for 10.0.0 release: It is ok to only push to use of Table instead of VectorSchemaRoot for single-batch processing?

Since the implementation is experimental I wouldn't 'push' people to use Table. I will update the docs which should let people know they have Table as an option and what the pros and cons are.
